src/OrleansConsulUtils/ConsulBasedMembershipTable.cs

```diff
@@ -0,0 +1,215 @@
+
+            return AssembleMembershipTableData(allSiloRegistrations);
+        }
+
+        public async Task<Boolean> InsertRow(MembershipEntry entry, TableVersion tableVersion)
+        {
+            try
+            {
+                //Ensure the KV for the entry does not already exist
+                var value = await _consulClient.KV.Get(ConsulSiloRegistration.ParseDeploymentSiloKey(_deploymentId, entry.SiloAddress));
+                if (value.Response != null)
+                {
+                    _logger.Verbose("ConsulMembershipProvider could not insert a registration for silo {0}, because one already exists.", entry.SiloAddress);
+                    return false;
+                }
+
+                var consulSiloKVEntry = ConsulSiloRegistration.FromMembershipEntry(_deploymentId, entry);
+                await _consulClient.KV.Put(consulSiloKVEntry.ToKVPair());
+
+                return true;
+            }
+            catch (Exception ex)
+            {
+                _logger.Verbose("ConsulMembershipProvider failed to insert registration for silo {0}; {1}.", entry.SiloAddress, ex);
+                return false;
+            }
+        }
+
+        public async Task<Boolean> UpdateRow(MembershipEntry entry, String etag, TableVersion tableVersion)
+        {
+            //Update Silo Liveness
+            try
+            {
+                var siloRegistration = await GetConsulSiloRegistration(entry.SiloAddress);
+                if (siloRegistration == null)
+                {
+                    _logger.Verbose("ConsulMembershipProvider could not find a registration for silo {0}.", entry.SiloAddress);
+                    return false;
+                }
+
+                siloRegistration.Status = entry.Status;
+                siloRegistration.IAmAliveTime = entry.IAmAliveTime;
+                siloRegistration.SuspectingSilos.AddRange(entry.SuspectTimes.Select(silo => new SuspectingSilo { Id = silo.Item1.ToParsableString(), Time = silo.Item2 }).ToList());
+
+                var updateKV = siloRegistration.ToKVPair();
+
+                //If the KV.CAS() call returns false then the update failed
+                var tryUpdate = await _consulClient.KV.CAS(updateKV);
+                if (!tryUpdate.Response)
+                {
+                    _logger.Verbose("ConsulMembershipProvider failed the CAS check when updating the registration for silo {0}.", entry.SiloAddress);
+                    return false;
+                }
+
+                return true;
+
+            }
+            catch (Exception ex)
+            {
+                _logger.Verbose("ConsulMembershipProvider failed to update the registration for silo {0}: {1}.", entry.SiloAddress, ex);
+                return false;
+            }
+        }
+
+        public async Task UpdateIAmAlive(MembershipEntry entry)
+        {
+            var siloRegistration = await GetConsulSiloRegistration(entry.SiloAddress);
```

Comment:
The MembershipEntry passed to UpdateIAmAlive is incomplete as Orleans expects the provider to support field-level update of the IAmAliveTime; because Consul can't do this, if we don't do a read before the update we will wipe out the port, hostname, starttime, etc. in the KV.  

This is possibly why the ZooKeeper implementation split the IAmAliveTime into a separate Key? We could do the same but then Update & InsertRow need to update 2 Keys and Consul does not support atomic updates.  Do you have any preference?